### PR TITLE
EKS smoke test

### PR DESCRIPTION
**Pull request by marktwallace**
This automates a very simple set of smoke tests to check if a zqd image can be deployed on EKS and works for basic distributed queries. An important feature of the smoke test is that it does not reserve exclusive resources, so it can run in parallel with itself.

The tests are simple, but the overall process is non-trivial. Since there is a lot to unpack here, I am interspersing comments in my own PR.

This adds two new jobs to what we have been running for PR's. The new jobs make extensive use of S3, so they will fail in ways that minio based testing does not.

The manual workflow dispatch dialog will be available only after this is pushed to master.

### CODE REVIEW

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:50]: I switched to deploying to a temporary namespace in order to avoid collisions when to PR's start builds close together. The smoke tests run in the temporary namespace and then we remove it.

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:94]: The next 5 commands are the smoke test. A basic search, using different numbers of workers, followed by a count() and a group-by query. we are not trying for extensive test coverage here: we just want to fail early if the basic distributed functionality is not working on K8s. (Similar queries are also run locally with ZTest.)

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:100]: The last two zapi queries are slow at the moment. We should fix that. Since they use 4 workers to do a full scan of the archive, they should be able to run at the same speed as the 4 worker search, about 10 seconds. But right now they take over 90 seconds. When we fix that these tests will finish sooner.

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:102]: I put in a small amount of diff-testing to ensure that the tests will fail when they need to. This is not intended to be a scalable pattern for diff testing -- if we want to have an extensive  test suite, then we should organize it similarly to ZTest. This pattern is OK for a few smoke tests, but it is not the way I would want to approach a longer running set of distributed tests.

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:107]: This is important -- if the namespace is not removed, we will run out of resources in our EKS cluster. We will need to keep an eye out to make sure that namespaces like "ci-310" are not hanging around in our zq-test cluster.

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:109]: The point of this second deploy to to make a tested version of zqd available in the ci-master namespace so we will have something relatively stable to use for dev testing like Brim UI development.

File: ppl/archive/multisource.go
Comment (marktwallace) [inline ppl/archive/multisource.go:134]: An error got thrown here a couple of times during test development, so we added a better message so we can track it down if it happens again. I would like to be able to reproduce it.

File: .github/workflows/cd.yaml
Comment (marktwallace) [inline .github/workflows/cd.yaml:80]: Note the new bucket used here: zq-ci-test is bucket with a short retention policy (3 days right now). A new directory is created for every build, so simultaneous builds will not collide while creating and querying archives. This creates S3 garbage fast but it will get cleared out by the retention policy.

File: .github/workflows/cd.yaml
Comment (philrz) [inline .github/workflows/cd.yaml:89]: This was discussed on a Zoom call, but I'm documenting it here for the record. I'm conditioned to react to these kinds of hard-coded sleep timers. @marktwallace mentioned that this has been passing fine, so we could defensibly let it ride. I've just been burned many times by these sorts of timers working fine at first, then one day infrastructure or automation decides to run abnormally slow, and now someone who doesn't know the automation as well has to react to a failure that didn't involve their code. It would seem ideal if we could instead poll some kind of "health endpoint" that's only reachable via the forwarded port, with the assumption that the health endpoint would only respond affirmatively if the service is truly 100% ready.

File: .github/workflows/cd.yaml
Comment (philrz) [inline .github/workflows/cd.yaml:125]: I know care was taken with the separate namespaces so concurrent runs shouldn't step on each other. However, at this point it seems to converge on a single namespace. Have we tested and/or know what will happen if two concurrent runs reach this step at the same time? I'm hopeful the underlying K8s operations are all atomic enough that one of them will consistently "win" and cancel the other out, as opposed to some kind of race causing deadlock or inconsistency, but figured I'd ask.

**Review by philrz — APPROVED** [review]
I expressed one preference for improvement and had one question. But overall I understand the essence of what's being done here, so I'm happy to see it go in. That said, I don't trust my ability to scrutinize all things K8s/AWS, so I'd recommend waiting on @alfred-landrum's feedback/approval before merging.

File: .github/workflows/cd.yaml
Comment (marktwallace) [reply .github/workflows/cd.yaml:147]: That is a good question: I think Helm will be alright with a second install starting during another install that is in-progress. We still should keep an eye on it, in case it can fail in some way I have not seen in my testing so far. Since it is our test environment, it is a good place to learn if and when collisions of this type can happen.

File: .github/workflows/cd.yaml
Comment (marktwallace) [reply .github/workflows/cd.yaml:111]: There are kubectl commands that can wait for objects to enter a "Ready" state, but being "ready" is not a guarantee that the port-forward will work, so the short sleep is a good idea in any case. A retry on the port-forward might be the easiest way to go.

File: .github/workflows/cd.yaml
Comment (alfred-landrum) [inline .github/workflows/cd.yaml:27]: For simplicity, if you need to use environment variables, like the `TODAY` and the `env` section, you could do so inside the run block:

```suggestion
        run: |
          set -x
          ZQD_DATA_URI=s3://zq-ci-test/$(date +'%y-%m-%d').${{ github.run_number }}
          ...
```

File: .github/workflows/cd.yaml
Comment (alfred-landrum) [reply .github/workflows/cd.yaml:117]: I filed #1934 to track follow up work to convert these to tagged ztests.

File: .github/workflows/cd.yaml
Comment (marktwallace) [reply .github/workflows/cd.yaml:24]: done

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:67]:
```suggestion
          export ZQD_DATA_URI=s3://zq-ci-test/$(date +%y-%m-%d).${{ github.run_number }}
```

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:22]: Nit: Set these in the `run` script. (Or not at all if they're only used once.)

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:64]: Nit: Set these in the `run` script. (Or not at all if they're only used once.)

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:95]: Nit: Set these in the `run` script. (Or not at all if they're only used once.)

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:111]:
```suggestion
          [[ 4 == $(curl http://localhost:8020/recruiter/listfree | jq -r ".workers | length") ]]     
```

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:124]:
```suggestion
        run: kubectl delete ns ci-${{ github.run_number }}
```

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:143]: Nit: Set these in the run script. (Or not at all if they're only used once.)

File: .github/workflows/cd.yaml
Comment (marktwallace) [reply .github/workflows/cd.yaml:95]: kubectl will only do one forward per line, unfortunately.

File: .github/workflows/cd.yaml
Comment (nwt) [inline .github/workflows/cd.yaml:28]: This file would be shorter (and clearer) if you used a single job rather than a sequence of jobs linked with `needs`. For instance, you'd only need one `uses: aws-actions/configure-aws-credentials@v1` step instead of four.

Comment (marktwallace) [reply .github/workflows/cd.yaml:28]: It is important that we can clearly distinguish between the jobs that fail and succeed. Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

Each job has a name that corresponds to its independent goal: clean, image build, deploy, test, master deploy. This code is optimized for operational clarity, as it should be. Making the code shorter is less important that the clarity in the operation of each job.

Comment (nwt) [reply .github/workflows/cd.yaml:28]: > Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

I don't think that's true, but the argument's tough to make in words, so let's merge this as is and explore later.

Comment (mattnibs) [reply .github/workflows/cd.yaml:28]: > Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

By this reasoning, you could say that we should break up every step in the "test" job (ie make fmt, make tidy, make test-unit, make test-system etc) into its own job would make it easier to identify where the failure occurred. But if we did this we would have so many jobs the list would be huge.  Fwiw I haven't had an issue with finding the reason for a failure in "test" job.

If this was called "eks-smoke test" and it failed while everything else passed, I would immediately have in my mind the parts of my pr that could have caused the failure. Having a small number of logically grouped together test jobs makes it easier to have a clear picture of what your ci test infrastructure looks like. If you a bunch of narrowly defined tests, you place a higher burden on institutional knowledge.

I don't want to hold up this pr, but I would prefer this was structured as a single job and moreover to keep an eye on the number of jobs we are creating.

Comment (marktwallace) [reply .github/workflows/cd.yaml:28]: The reasoning is different from what you describe. This is not a single test. It is a collection of the most important external integration points with the deployment technologies we are using. If the Docker build fails, that is a much more significant failure than a unit test: it means our code is not be being built on our selected deployment platform. The deploy step failing means that our AWS infrastructure cannot accept what we built, and EKS may be misconfigured, or our K8s yaml is broken. This is a very different type of failure than a functional test. The test step in this will contain an extensive set of distributed tests, reflecting a subset of our local ZTests, and need not be broken into multiple jobs.

File: .github/workflows/cd.yaml
Comment (mattnibs) [inline .github/workflows/cd.yaml:7]: This should be removed no?

Comment (marktwallace) [reply .github/workflows/cd.yaml:7]: Yes, I should have taken it out. However, it does no harm, so not urgent.

**Pull request merged**
